### DISCUSSION TITLE
Get state value

### DISCUSSION THREAD

**Discussion by gohryt** [xania/view#27] [Q&A] [answered]
Hello, i have a simple component which code is listed below:
```
import { State, useState } from "xania";

import style from "./Head.module.css";

import iconLogo from "./icons/head/logo.svg";
import iconTON from "./icons/head/ton.svg";
import iconShare from "./icons/head/share.svg";

export const Head = async (): Promise<JSX.Element> => {
	const flip: State<boolean> = useState(false);

	const classes: string[] = [style.logo, flip.when(true, style.flip, "")]

	return (<div className={style.head}>
		<div className={classes.join(" ")} pointerdown={flip.update((flip) => !flip)}>
			<img src={iconLogo} />
		</div>
		<div className={style.title}>
			<img src={iconTON} />
			<span className={style.titleText}>TON</span>
		</div>
		<div className={style.share}>
			<img src={iconShare} />
		</div>
	</div >);
}
```
I want its logo rotation by 180deg on click by adding class "flip" from css module, but i can't check it like this
```flip ? style.flip : ""``` or ```flip.when(true, style.flip, "")``` or something else because state becomes strange class, all state methods returns strange classes and all of it has no member like value or so.
How can i do this?

**Comment by gohryt** [comment]
i do it now by
```
import { State, useState } from "xania";

import style from "./Head.module.css";

import iconLogo from "./icons/head/logo.svg";
import iconTON from "./icons/head/ton.svg";
import iconShare from "./icons/head/share.svg";

export const Head = async (): Promise<JSX.Element> => {
	const flip = useState<boolean | null>(null);

	const classList = flip.map((flip) => {
		switch (flip) {
			case true:
				return `${style.logo} ${style.flip}`
			case false:
				return `${style.logo} ${style.back}`
			default:
				return style.logo
		}
	});

	return (<div className={style.head}>
		<div className={classList} pointerdown={flip.update((flip) => !flip)}>
			<img src={iconLogo} />
		</div>
		<div className={style.title}>
			<img src={iconTON} />
			<span className={style.titleText}>TON</span>
		</div>
		<div className={style.share}>
			<img src={iconShare} />
		</div>
	</div >);
}
```
but it seems like it may be only temporary solution because i can have multiple state based conditions.

**Reply by xania** [reply, marked as answer]
@gohryt ideomatic way to pass multiple class values is to use array, you can mix reactive with constant string in the array, you can even put an array inside an array (inside an array, inside an array....) and xania will flatten that for you (If not then this will be coming soon).

Try this and let me know
```
<div class={[style.logo, flip.when(true, style.flip, style.back)]} pointerdown={flip.update((flip) => !flip)}>
   <img src={iconLogo} />
</div>
```

**Reply by gohryt** [reply]
this works, thanks you for your great job, the reason i didn't do it is
![image](https://github.com/xania/view/assets/38947781/e947175a-898f-4c21-9d9c-97cf9ddb5a42)

**Reply by gohryt** [reply]
The only remaining thing with status is a problem when i need to update two states in one action listener:
![image](https://github.com/xania/view/assets/38947781/538f455f-96d0-421b-8f43-e9b6b8997d78)
Array solution (as u can see above) is not working.
the one listener and code as below
![image](https://github.com/xania/view/assets/38947781/7fbec1c9-20c5-4c51-91c6-46da7b09ab93)
is not working too.
I tried to understand how it works internally with debugger and it seems like runtime has no option in switch for arrays, and as i understand command only executes when returned to runtime.
The discussion about accessing state values is very important because we can have parent component state and current component state and in some cases may need in their combination:
Parent (some layer, header as example) - theme state
Current (button as example) - condition state (button may be disabled or something else)
In this case we can have (also as example) this combinations of states:
theme light + button active = green
theme light + button disabled = red
theme dark + button active = white
theme dark + button disabled = grey
There wee shoud have an option to do it.

**Reply by xania** [reply]
the line `attached.update(...)` in body of `listener` has no effect if it is not returned to the runtime. 

There is however an issue with binding array to click event, a workaround is to bind to a function that returns an array. see below a working example

```jsx
import { render, useState } from "xania";
import style from "./style.module.scss";

render(<App />, document.getElementById("app")!);

function App() {
  const dark = useState(false);
  const label = useState("Click me");

  return (
    <div>
      <button
        class={["head", dark.when(true, style.light, style.dark)]}
        click={() => [dark.update((v) => !v), label.update((s) => s + "!")]}
      >
        {label}
      </button>
    </div>
  );
}
```

**Reply by xania** [reply]
Another issue is type checking of `className` attribute, use `class` to get rid of the type errors

**Comment by xania** [comment]
Hi, Quick observation, you shouldn't call join(" ") because that will break reactivity 
I will look at ur question more closely later on but this is something u could try

**Comment by xania** [comment]
@gohryt on ur comment about not able to access the value of the state, that's correct by design, as you already figured out, you can use state method like `.map` to access the value. Another way is to use `update` to access all values of current _scope_ as follows

```
import { update } from "xania"

function MyComp() {
  const s = useState(1);
  const onClick = update(async function (scope) {
    var x = await scope.get(s);
    return s.update(x! + 1);
  });
   return (
       <button click={onClick} >click me!</button>
   );
}

```
